core/chain_pow_test.go

```diff
@@ -0,0 +1,232 @@
+func (pow delayedPow) Turbo(bool)                  {}
+
+// Tests that simple POW verification works, for both good and bad blocks.
+func TestPowVerification(t *testing.T) {
+	// Create a simple chain to verify
+	var (
+		testdb, _ = ethdb.NewMemDatabase()
+		genesis   = GenesisBlockForTesting(testdb, common.Address{}, new(big.Int))
+		blocks    = GenerateChain(genesis, testdb, 8, nil)
+	)
+	headers := make([]*types.Header, len(blocks))
+	for i, block := range blocks {
+		headers[i] = block.Header()
+	}
+	// Run the POW checker for blocks one-by-one, checking for both valid and invalid nonces
+	for i := 0; i < len(blocks); i++ {
+		for j, full := range []bool{true, false} {
+			for k, valid := range []bool{true, false} {
+				var results <-chan nonceCheckResult
+
+				switch {
+				case full && valid:
+					_, results = verifyBlockNonces(FakePow{}, []*types.Block{blocks[i]})
+				case full && !valid:
+					_, results = verifyBlockNonces(failPow{blocks[i].NumberU64()}, []*types.Block{blocks[i]})
+				case !full && valid:
+					_, results = verifyHeaderNonces(FakePow{}, []*types.Header{headers[i]})
+				case !full && !valid:
+					_, results = verifyHeaderNonces(failPow{headers[i].Number.Uint64()}, []*types.Header{headers[i]})
+				}
+				// Wait for the verification result
+				select {
+				case result := <-results:
+					if result.index != 0 {
+						t.Errorf("test %d.%d.%d: invalid index: have %d, want 0", i, j, k, result.index)
+					}
+					if result.valid != valid {
+						t.Errorf("test %d.%d.%d: validity mismatch: have %v, want %v", i, j, k, result.valid, valid)
+					}
+				case <-time.After(time.Second):
+					t.Fatalf("test %d.%d.%d: verification timeout", i, j, k)
+				}
+				// Make sure no more data is returned
+				select {
+				case result := <-results:
+					t.Fatalf("test %d.%d.%d: unexpected result returned: %v", i, j, k, result)
+				case <-time.After(25 * time.Millisecond):
+				}
+			}
+		}
+	}
+}
+
+// Tests that concurrent POW verification works, for both good and bad blocks.
+func TestPowConcurrentVerification2(t *testing.T)  { testPowConcurrentVerification(t, 2) }
+func TestPowConcurrentVerification8(t *testing.T)  { testPowConcurrentVerification(t, 8) }
+func TestPowConcurrentVerification32(t *testing.T) { testPowConcurrentVerification(t, 32) }
+
+func testPowConcurrentVerification(t *testing.T, threads int) {
+	// Create a simple chain to verify
+	var (
+		testdb, _ = ethdb.NewMemDatabase()
+		genesis   = GenesisBlockForTesting(testdb, common.Address{}, new(big.Int))
+		blocks    = GenerateChain(genesis, testdb, 8, nil)
+	)
+	headers := make([]*types.Header, len(blocks))
+	for i, block := range blocks {
+		headers[i] = block.Header()
+	}
+	// Set the number of threads to verify on
+	old := runtime.GOMAXPROCS(threads)
+	defer runtime.GOMAXPROCS(old)
+
+	// Run the POW checker for blocks one-by-one, checking for both valid and invalid nonces
+	for i, full := range []bool{true, false} {
```

Comment:
Same here, could read easier with either a func or proper function following each of the 4 calls to ```verifyBlockNonces```